usd/src/usd_parser/USDTransforms.cc

```diff
@@ -0,0 +1,395 @@
+}
+
+//////////////////////////////////////////////////
+const ignition::math::Vector3d UDSTransforms::Scale() const
+{
+  return this->dataPtr->scale;
+}
+
+//////////////////////////////////////////////////
+const std::vector<ignition::math::Quaterniond>
+  UDSTransforms::Rotations() const
+{
+  return this->dataPtr->q;
+}
+
+//////////////////////////////////////////////////
+void UDSTransforms::Translate(
+  const ignition::math::Vector3d &_translate)
+{
+  this->dataPtr->translate = _translate;
+}
+
+//////////////////////////////////////////////////
+void UDSTransforms::Scale(
+  const ignition::math::Vector3d &_scale)
+{
+  this->dataPtr->scale = _scale;
+}
+
+//////////////////////////////////////////////////
+void UDSTransforms::AddRotation(
+  const ignition::math::Quaterniond &_q)
+{
+  this->dataPtr->q.push_back(_q);
+}
+
+//////////////////////////////////////////////////
+bool UDSTransforms::RotationZYX() const
+{
+  return this->dataPtr->isRotationZYX;
+}
+
+//////////////////////////////////////////////////
+bool UDSTransforms::RotationXYZ() const
+{
+  return this->dataPtr->isRotationXYZ;
+}
+
+//////////////////////////////////////////////////
+bool UDSTransforms::Rotation() const
+{
+  return this->dataPtr->isRotation;
+}
+
+//////////////////////////////////////////////////
+void UDSTransforms::RotationZYX(bool _rotationZYX)
+{
+  this->dataPtr->isRotationZYX = _rotationZYX;
+}
+
+//////////////////////////////////////////////////
+void UDSTransforms::RotationXYZ(bool _rotationXYZ)
+{
+  this->dataPtr->isRotationXYZ = _rotationXYZ;
+}
+
+//////////////////////////////////////////////////
+void UDSTransforms::Rotation(bool _rotation)
+{
+  this->dataPtr->isRotation = _rotation;
+}
+
+//////////////////////////////////////////////////
+void GetAllTransforms(
+  const pxr::UsdPrim &_prim,
+  const USDData &_usdData,
+  std::vector<ignition::math::Pose3d> &_tfs,
+  ignition::math::Vector3d &_scale,
+  const std::string &_schemaToStop)
+{
+  pxr::UsdPrim parent = _prim;
+  double metersPerUnit = 1.0;
+  std::string upAxis = "Y";
+
+  // this assumes that there can only be one stage
+  const auto stageData = _usdData.FindStage(parent.GetPath().GetName());
+  if (stageData.second != nullptr) {
+    metersPerUnit = stageData.second->MetersPerUnit();
+    upAxis = stageData.second->UpAxis();
+  }
+
+  while (parent)
+  {
+    if (pxr::TfStringify(parent.GetPath()) == _schemaToStop)
+    {
+      return;
+    }
+
+    UDSTransforms t = ParseUSDTransform(parent);
+
+    ignition::math::Pose3d pose;
+    _scale *= t.Scale();
+
+    pose.Pos() = t.Translate() * metersPerUnit;
+    // scaling is lost when we convert to pose, so we pre-scale the
+    // translation to make them match the scaled values.
+    if (!_tfs.empty()) {
+      auto& child = _tfs.back();
+      child.Pos().Set(
+        child.Pos().X() * t.Scale()[0],
+        child.Pos().Y() * t.Scale()[1],
+        child.Pos().Z() * t.Scale()[2]);
+    }
```

Comment:
This part was done by @koonpeng maybe he can add something here

Reply:
iirc, the last pose's scale is being used as the value for the model's scale.

Reply:
Thanks for the explanation. It sounds like we are assuming that the scale is defined in a parent prim of whatever we are transforming, but is this a valid assumption? For example, what if in the example you just gave, `TransformChild` has `translate[10,0,0]` and also has `scale[10,0,0]`? In this case, I believe the result should be `[100,0,0]`, but we would instead have this saved as `[10,0,0]` because we looked for the scale in the parent instead of the child.

Reply:
Checking the code again, the pose and scaling is computed by multiplying each children transform. So given, `TransformChild` with `translate[10,0,0], scale[2,0,0]` and `TransformParent` with `translate[20,0,0],scale[10,0,0]`. The final pose will be

ParentPose * ChildPose = translate[30,0,0] (or is it Child * Parent? Can't remeber the order of operations in pose multipliciations)
ChildScale * ParentScale = scale[20,0,0]

Given a point [0,0,0], that will end up as [30,0,0] which is wrong. What should have happened is TransformChild moved it to [10,0,0], then TransformParent moves it to [120,0,0]. The scaling applied on the intermediate point is lost because we separated the calculation of the pose and scaling.